src/devices/cpu/arcompact/arcompact.h

```diff
@@ -20,69 +18,93 @@ enum
 
 };
 
-#define ARCOMPACT_RETTYPE uint32_t
-#define OPS_32 uint32_t op
-#define OPS_16 uint16_t op
-#define PARAMS op
 #define LIMM_REG 62
-#define ARCOMPACT_OPERATION ((op & 0xf800) >> 11)
-
-
-#define ARCOMPACT_HANDLER04_P11_TYPE(name) \
-ARCOMPACT_RETTYPE arcompact_handle##name##_p11(OPS_32) \
-{ \
-	int M = (op & 0x00000020) >> 5; \
-		\
-	switch (M) \
-	{ \
-		case 0x00: return arcompact_handle##name##_p11_m0(PARAMS); \
-		case 0x01: return arcompact_handle##name##_p11_m1(PARAMS); \
-	} \
-		\
-	return 0; \
-}
-#define ARCOMPACT_HANDLER04_TYPE(name) \
-ARCOMPACT_RETTYPE arcompact_handle##name(OPS_32) \
-{ \
-	int p = (op & 0x00c00000) >> 22; \
-	\
-	switch (p) \
-	{ \
-		case 0x00: return arcompact_handle##name##_p00(PARAMS); \
-		case 0x01: return arcompact_handle##name##_p01(PARAMS); \
-		case 0x02: return arcompact_handle##name##_p10(PARAMS); \
-		case 0x03: return arcompact_handle##name##_p11(PARAMS); \
-	} \
-	\
-	return 0; \
-}
-
-#define ARCOMPACT_HANDLER04_TYPE_PM(name) \
-	ARCOMPACT_RETTYPE arcompact_handle##name##_p00(OPS_32); \
-	ARCOMPACT_RETTYPE arcompact_handle##name##_p01(OPS_32); \
-	ARCOMPACT_RETTYPE arcompact_handle##name##_p10(OPS_32); \
-	ARCOMPACT_RETTYPE arcompact_handle##name##_p11_m0(OPS_32); \
-	ARCOMPACT_RETTYPE arcompact_handle##name##_p11_m1(OPS_32); \
-	ARCOMPACT_HANDLER04_P11_TYPE(name); \
-	ARCOMPACT_HANDLER04_TYPE(name);
+
+#define REG_BLINK (0x1f) // r31
+#define REG_SP (0x1c) // r28
+#define REG_ILINK1 (0x1d) // r29
+#define REG_ILINK2 (0x1e) // r30
+#define REG_LP_COUNT (0x3c) // r60
+
+#define ARCOMPACT_LOGGING 1
+
+#define arcompact_fatal if (ARCOMPACT_LOGGING) fatalerror
+#define arcompact_log if (ARCOMPACT_LOGGING) fatalerror
+
+
+#define GET_01_01_01_BRANCH_ADDR \
+	int32_t address = (op & 0x00fe0000) >> 17; \
+	address |= ((op & 0x00008000) >> 15) << 7; \
+	if (address & 0x80) address = -0x80 + (address & 0x7f);
+
+#define GROUP_0e_GET_h \
+	h =  ((op & 0x0007) << 3); \
+	h |= ((op & 0x00e0) >> 5);
+#define COMMON32_GET_breg \
+	int b_temp = (op & 0x07000000) >> 24; \
+	int B_temp = (op & 0x00007000) >> 12; \
+	int breg = b_temp | (B_temp << 3);
+#define COMMON32_GET_creg \
+	int creg = (op & 0x00000fc0) >> 6;
+#define COMMON32_GET_u6 \
+	int u = (op & 0x00000fc0) >> 6;
+#define COMMON32_GET_areg \
+	int areg = (op & 0x0000003f) >> 0;
+#define COMMON32_GET_areg_reserved \
+	int ares = (op & 0x0000003f) >> 0;
+#define COMMON32_GET_F \
+	int F = (op & 0x00008000) >> 15;
+#define COMMON32_GET_p \
+	int p = (op & 0x00c00000) >> 22;
```

Comment:
This kind of stuff really shouldn’t leak downstream – it should go in a local `.ipp` file that can be included by all the implementation source files, along with inline member implementations etc.